File: backend/utils/llm_client.py
"""Unified cloud LLM client using the OpenAI SDK.

Replaces ollama_client.py.  Provides the same return format so that
agent code requires minimal changes.
"""

from __future__ import annotations

import json
import logging
import re
from typing import Any, AsyncGenerator

import httpx
from openai import AsyncOpenAI

from backend.config import MODEL_TEMPERATURE

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def parse_thinking_tags(text: str) -> tuple[str, str]:
    """Separate <think>...</think> content from the final answer.

    Returns (thinking, answer).
    """
    match = re.search(r"<think>(.*?)</think>", text, re.DOTALL)
    if match:
        thinking = match.group(1).strip()
        answer = text[: match.start()] + text[match.end() :]
        return thinking, answer.strip()
    return "", text.strip()


def _prep_messages(messages: list[dict[str, Any]]) -> list[dict[str, Any]]:
    """Prepare messages for the OpenAI API.

    - Convert tool_calls arguments from dict to JSON string.
    - Ensure tool response messages have a tool_call_id.
    """
    prepped: list[dict[str, Any]] = []
    # Collect tool_call ids from assistant messages so we can match them to
    # subsequent tool-role messages.
    pending_tool_call_ids: list[str] = []

    for msg in messages:
        m = dict(msg)  # shallow copy
        role = m.get("role", "")

        if role == "assistant" and "tool_calls" in m and m["tool_calls"]:
            tcs = []
            for tc in m["tool_calls"]:
                tc_copy = dict(tc)
                func = dict(tc_copy.get("function", {}))
                # Ensure arguments is a JSON string
                if isinstance(func.get("arguments"), dict):
                    func["arguments"] = json.dumps(func["arguments"], ensure_ascii=False)
                tc_copy["function"] = func
                # Generate an id if missing
                if "id" not in tc_copy or not tc_copy["id"]:
                    tc_copy["id"] = f"call_{len(prepped)}_{len(tcs)}"
                tcs.append(tc_copy)
                pending_tool_call_ids.append(tc_copy["id"])
            m["tool_calls"] = tcs
            # OpenAI requires content to be null (not empty string) when tool_calls present
            if not m.get("content"):
                m["content"] = None

        elif role == "tool":
            # Assign tool_call_id from the pending queue
            if "tool_call_id" not in m or not m["tool_call_id"]:
                if pending_tool_call_ids:
                    m["tool_call_id"] = pending_tool_call_ids.pop(0)
                else:
                    m["tool_call_id"] = f"call_fallback_{len(prepped)}"

        prepped.append(m)
    return prepped


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------

class LLMClient:
    """Async OpenAI-compatible LLM client."""

    def __init__(self) -> None:
        self._client: AsyncOpenAI | None = None
        self._model: str = ""
        self._provider_id: str = ""
        self._thinking_field: str = ""  # e.g. "reasoning_content" for DeepSeek

    # -- Configuration -------------------------------------------------------

    def configure(
        self,
        *,
        api_key: str,
        base_url: str,
        model: str,
        provider_id: str = "",
        thinking_field: str = "",
    ) -> None:
        """Create / recreate the underlying AsyncOpenAI instance."""
        # BUG-8 fix: bypass Windows system proxy
        http_client = httpx.AsyncClient(trust_env=False)
        self._client = AsyncOpenAI(
            api_key=api_key,
            base_url=base_url.rstrip("/"),
            http_client=http_client,
        )
        self._model = model
        self._provider_id = provider_id
        self._thinking_field = thinking_field
        logger.info(
            "LLMClient configured: provider=%s model=%s base_url=%s",
            provider_id, model, base_url,
        )

    @property
    def is_configured(self) -> bool:
        return self._client is not None and bool(self._model)

    @property
    def model(self) -> str:
        return self._model

    @property
    def provider_id(self) -> str:
        return self._provider_id

    # -- Chat (non-streaming) ------------------------------------------------

    async def chat(
        self,
        messages: list[dict[str, Any]],
        *,
        temperature: float = MODEL_TEMPERATURE,
        max_tokens: int | None = None,
        think: bool | None = None,
    ) -> dict[str, Any]:
        """Non-streaming chat. Returns Ollama-compatible format:

        {"message": {"role": "assistant", "content": "...", "thinking": "..."}}
        """
        if not self._client:
            raise RuntimeError("LLM client not configured")

        prepped = _prep_messages(messages)
        kwargs: dict[str, Any] = {
            "model": self._model,
            "messages": prepped,
            "temperature": temperature,
        }
        if max_tokens is not None:
            kwargs["max_tokens"] = max_tokens

        logger.debug("LLM chat: model=%s messages=%d", self._model, len(prepped))
        resp = await self._client.chat.completions.create(**kwargs)

        choice = resp.choices[0]
        content = choice.message.content or ""
        thinking = ""

        # Extract thinking from provider-specific field
        if self._thinking_field:
            thinking = getattr(choice.message, self._thinking_field, "") or ""

        # Fallback: parse <think> tags from content
        if not thinking and content:
            thinking, content = parse_thinking_tags(content)

        return {
            "message": {
                "role": "assistant",
                "content": content,
                "thinking": thinking,
            }
        }

    # -- Chat (streaming) ----------------------------------------------------

    async def chat_stream(
        self,
        messages: list[dict[str, Any]],
        *,
        temperature: float = MODEL_TEMPERATURE,
        max_tokens: int | None = None,
    ) -> AsyncGenerator[dict[str, str], None]:
        """Streaming chat. Yields Ollama-compatible chunks:

        {"thinking": "token"} / {"content": "token"} / {"done": "true"}
        """
        if not self._client:
            raise RuntimeError("LLM client not configured")

        prepped = _prep_messages(messages)
        kwargs: dict[str, Any] = {
            "model": self._model,
            "messages": prepped,
            "temperature": temperature,
            "stream": True,
        }
        if max_tokens is not None:
            kwargs["max_tokens"] = max_tokens

        logger.debug("LLM chat_stream: model=%s messages=%d", self._model, len(prepped))

        # Track whether we're inside <think> tags for providers that embed them
        in_think_tag = False
        content_buffer = ""

        stream = await self._client.chat.completions.create(**kwargs)
        async for chunk in stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            if delta is None:
                continue

            # Provider-specific thinking field (DeepSeek reasoner)
            if self._thinking_field:
                reasoning = getattr(delta, self._thinking_field, None)
                if reasoning:
                    yield {"thinking": reasoning}

            text = delta.content or ""
            if not text:
                continue

            # Parse <think> tags in-stream for providers that embed them
            if not self._thinking_field:
                # Process character by character for tag detection
                for char in text:
                    content_buffer += char
                    if content_buffer.endswith("<think>"):
                        in_think_tag = True
                        # Remove the tag from buffer
                        content_buffer = content_buffer[:-7]
                        if content_buffer:
                            yield {"content": content_buffer}
                            content_buffer = ""
                    elif content_buffer.endswith("</think>"):
                        in_think_tag = False
                        # Remove the tag from buffer and emit as thinking
                        content_buffer = content_buffer[:-8]
                        if content_buffer:
                            yield {"thinking": content_buffer}
                            content_buffer = ""
                    elif len(content_buffer) > 20:
                        # Flush buffer
                        if in_think_tag:
                            yield {"thinking": content_buffer}
                        else:
                            yield {"content": content_buffer}
                        content_buffer = ""
            else:
                yield {"content": text}

        # Flush remaining buffer
        if content_buffer:
            if in_think_tag:
                yield {"thinking": content_buffer}
            else:
                yield {"content": content_buffer}

        yield {"done": "true"}

    # -- Chat with tools (non-streaming) ------------------------------------

    async def chat_with_tools(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
        *,
        temperature: float = MODEL_TEMPERATURE,
        max_tokens: int | None = None,
    ) -> dict[str, Any]:
        """Non-streaming chat with tool calling. Returns Ollama-compatible format:

        {"message": {"role": "assistant", "content": "...", "thinking": "...",
                      "tool_calls": [{"id": "...", "function": {"name": "...", "arguments": {...}}}]}}
        """
        if not self._client:
            raise RuntimeError("LLM client not configured")

        prepped = _prep_messages(messages)
        kwargs: dict[str, Any] = {
            "model": self._model,
            "messages": prepped,
            "temperature": temperature,
        }
        if max_tokens is not None:
            kwargs["max_tokens"] = max_tokens
        if tools:
            kwargs["tools"] = tools

        logger.debug(
            "LLM chat_with_tools: model=%s tools=%d messages=%d",
            self._model, len(tools), len(prepped),
        )
        resp = await self._client.chat.completions.create(**kwargs)

        choice = resp.choices[0]
        content = choice.message.content or ""
        thinking = ""

        # Extract thinking
        if self._thinking_field:
            thinking = getattr(choice.message, self._thinking_field, "") or ""
        if not thinking and content:
            thinking, content = parse_thinking_tags(content)

        # Convert tool calls to Ollama-compatible format
        tool_calls_out: list[dict[str, Any]] = []
        if choice.message.tool_calls:
            for tc in choice.message.tool_calls:
                # Parse arguments from JSON string to dict
                try:
                    args = json.loads(tc.function.arguments)
                except (json.JSONDecodeError, TypeError):
                    args = {}
                tool_calls_out.append({
                    "id": tc.id,
                    "function": {
                        "name": tc.function.name,
                        "arguments": args,
                    },
                })

        result: dict[str, Any] = {
            "message": {
                "role": "assistant",
                "content": content,
                "thinking": thinking,
            }
        }
        if tool_calls_out:
            result["message"]["tool_calls"] = tool_calls_out

        return result

    # -- Health check --------------------------------------------------------

    async def check_health(self) -> tuple[bool, bool]:
        """Check API reachability and model availability.

        Returns (api_ok, model_ok).  Both True if a simple completion succeeds.
        """
        if not self._client:
            return False, False
        try:
            resp = await self._client.chat.completions.create(
                model=self._model,
                messages=[{"role": "user", "content": "ping"}],
                max_tokens=1,
            )
            return True, bool(resp.choices)
        except Exception as e:
            logger.warning("LLM health check failed: %s", e)
            # Distinguish API-level vs model-level errors
            err_str = str(e).lower()
            if "model" in err_str or "not found" in err_str:
                return True, False
            return False, False


# Singleton
llm_client = LLMClient()

File: frontend/src/core/llm-client.ts
import { getProvider } from './providers'

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ChatMessage {
  role: 'system' | 'user' | 'assistant' | 'tool'
  content: string | null
  thinking?: string | null
  tool_calls?: ToolCall[]
  tool_call_id?: string
  name?: string
}

export interface ToolCall {
  id: string
  type: 'function'
  function: {
    name: string
    arguments: Record<string, any>
  }
}

export interface ToolDefinition {
  type: 'function'
  function: {
    name: string
    description: string
    parameters: Record<string, any>
  }
}

export interface ChatResponse {
  message: {
    role: string
    content: string | null
    thinking: string | null
  }
}

export interface ToolResponse {
  message: {
    role: string
    content: string | null
    thinking: string | null
    tool_calls: ToolCall[] | null
  }
}

export type StreamChunk =
  | { thinking: string }
  | { content: string }
  | { done: 'true' }

export interface HealthStatus {
  apiOk: boolean
  modelOk: boolean
}

interface LLMConfig {
  apiKey: string
  baseUrl: string
  model: string
  providerId: string
  thinkingField: string
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Extract thinking content from `<think>...</think>` tags and return
 * the cleaned content plus extracted thinking.
 */
function extractThinkTags(text: string): { content: string; thinking: string } {
  const regex = /<think>([\s\S]*?)<\/think>/g
  const thinkParts: string[] = []
  const cleaned = text.replace(regex, (_match, inner) => {
    thinkParts.push(inner.trim())
    return ''
  })
  return {
    content: cleaned.trim(),
    thinking: thinkParts.join('\n'),
  }
}

/**
 * Prepare messages for the API: serialise tool_call arguments to JSON strings
 * and ensure tool messages carry a tool_call_id.
 */
function prepMessages(messages: ChatMessage[]): Record<string, any>[] {
  return messages.map((msg) => {
    const out: Record<string, any> = {
      role: msg.role,
      content: msg.content,
    }

    if (msg.role === 'tool') {
      if (msg.tool_call_id) out.tool_call_id = msg.tool_call_id
      if (msg.name) out.name = msg.name
    }

    if (msg.tool_calls && msg.tool_calls.length > 0) {
      out.tool_calls = msg.tool_calls.map((tc) => ({
        id: tc.id,
        type: tc.type,
        function: {
          name: tc.function.name,
          arguments:
            typeof tc.function.arguments === 'string'
              ? tc.function.arguments
              : JSON.stringify(tc.function.arguments),
        },
      }))
    }

    return out
  })
}

// ---------------------------------------------------------------------------
// LLMClient
// ---------------------------------------------------------------------------

export class LLMClient {
  private config: LLMConfig = {
    apiKey: '',
    baseUrl: '',
    model: '',
    providerId: '',
    thinkingField: '',
  }

  private defaultTemperature = 0.6

  // ── Configuration ──────────────────────────────────────────────────────

  configure(opts: {
    apiKey: string
    baseUrl: string
    model: string
    providerId?: string
    thinkingField?: string
  }): void {
    this.config.apiKey = opts.apiKey
    this.config.baseUrl = opts.baseUrl.replace(/\/+$/, '')
    this.config.model = opts.model
    this.config.providerId = opts.providerId ?? ''

    if (opts.thinkingField !== undefined) {
      this.config.thinkingField = opts.thinkingField
    } else if (opts.providerId) {
      const provider = getProvider(opts.providerId)
      this.config.thinkingField = provider?.thinking_field ?? ''
    }
  }

  get isConfigured(): boolean {
    return !!(this.config.apiKey && this.config.baseUrl && this.config.model)
  }

  // ── Private fetch wrapper ──────────────────────────────────────────────

  private get endpoint(): string {
    return `${this.config.baseUrl}/chat/completions`
  }

  private buildHeaders(): Record<string, string> {
    return {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.config.apiKey}`,
    }
  }

  private async request(
    body: Record<string, any>,
    stream: false,
  ): Promise<any>
  private async request(
    body: Record<string, any>,
    stream: true,
  ): Promise<ReadableStream<Uint8Array>>
  private async request(
    body: Record<string, any>,
    stream: boolean,
  ): Promise<any | ReadableStream<Uint8Array>> {
    const response = await fetch(this.endpoint, {
      method: 'POST',
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
    })

    if (!response.ok) {
      let detail = ''
      try {
        const errJson = await response.json()
        detail = errJson.error?.message ?? JSON.stringify(errJson)
      } catch {
        detail = await response.text().catch(() => '')
      }
      throw new Error(
        `LLM API 错误 (${response.status}): ${detail || response.statusText}`,
      )
    }

    if (stream) {
      if (!response.body) throw new Error('响应流不可用')
      return response.body
    }
    return response.json()
  }

  // ── Thinking extraction helper ─────────────────────────────────────────

  private extractThinking(choice: any): { content: string | null; thinking: string | null } {
    const msg = choice.message ?? choice.delta ?? {}
    let content: string | null = msg.content ?? null
    let thinking: string | null = null

    // 1) Check provider-specific thinking field (e.g. reasoning_content)
    if (this.config.thinkingField && msg[this.config.thinkingField]) {
      thinking = msg[this.config.thinkingField]
    }

    // 2) Fallback: parse <think>...</think> from content
    if (!thinking && content) {
      const extracted = extractThinkTags(content)
      if (extracted.thinking) {
        thinking = extracted.thinking
        content = extracted.content || null
      }
    }

    return { content, thinking }
  }

  // ── chat() — non-streaming ─────────────────────────────────────────────

  async chat(
    messages: ChatMessage[],
    opts?: { temperature?: number },
  ): Promise<ChatResponse> {
    const body: Record<string, any> = {
      model: this.config.model,
      messages: prepMessages(messages),
      temperature: opts?.temperature ?? this.defaultTemperature,
      stream: false,
    }

    const data = await this.request(body, false)
    const choice = data.choices?.[0]
    if (!choice) throw new Error('LLM 返回空响应')

    const { content, thinking } = this.extractThinking(choice)

    return {
      message: {
        role: choice.message.role,
        content,
        thinking,
      },
    }
  }

  // ── chatStream() — SSE streaming ───────────────────────────────────────

  async *chatStream(
    messages: ChatMessage[],
    opts?: { temperature?: number },
  ): AsyncGenerator<StreamChunk> {
    const body: Record<string, any> = {
      model: this.config.model,
      messages: prepMessages(messages),
      temperature: opts?.temperature ?? this.defaultTemperature,
      stream: true,
    }

    const stream = await this.request(body, true)
    const reader = stream.getReader()
    const decoder = new TextDecoder()
    let buffer = ''

    // Track whether we're inside a <think> block while streaming
    let inThinkTag = false
    let thinkBuffer = ''

    try {
      while (true) {
        const { done, value } = await reader.read()
        if (done) break

        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        // Keep the last (potentially incomplete) line in the buffer
        buffer = lines.pop() ?? ''

        for (const line of lines) {
          const trimmed = line.trim()
          if (!trimmed || !trimmed.startsWith('data:')) continue

          const payload = trimmed.slice(5).trim()
          if (payload === '[DONE]') {
            yield { done: 'true' }
            return
          }

          let parsed: any
          try {
            parsed = JSON.parse(payload)
          } catch {
            continue
          }

          const delta = parsed.choices?.[0]?.delta
          if (!delta) continue

          // Provider-specific thinking field (e.g. reasoning_content)
          if (this.config.thinkingField && delta[this.config.thinkingField]) {
            yield { thinking: delta[this.config.thinkingField] }
          }

          if (delta.content) {
            // Handle <think> tags that may appear mid-stream
            let text = delta.content as string

            while (text.length > 0) {
              if (inThinkTag) {
                const closeIdx = text.indexOf('</think>')
                if (closeIdx === -1) {
                  thinkBuffer += text
                  text = ''
                } else {
                  thinkBuffer += text.slice(0, closeIdx)
                  if (thinkBuffer) yield { thinking: thinkBuffer }
                  thinkBuffer = ''
                  inThinkTag = false
                  text = text.slice(closeIdx + 8) // skip </think>
                }
              } else {
                const openIdx = text.indexOf('<think>')
                if (openIdx === -1) {
                  yield { content: text }
                  text = ''
                } else {
                  if (openIdx > 0) yield { content: text.slice(0, openIdx) }
                  inThinkTag = true
                  text = text.slice(openIdx + 7) // skip <think>
                }
              }
            }
          }
        }
      }

      // Flush remaining think buffer
      if (thinkBuffer) yield { thinking: thinkBuffer }
    } finally {
      reader.releaseLock()
    }

    yield { done: 'true' }
  }

  // ── chatWithTools() — tool calling ─────────────────────────────────────

  async chatWithTools(
    messages: ChatMessage[],
    tools: ToolDefinition[],
    opts?: { temperature?: number },
  ): Promise<ToolResponse> {
    const body: Record<string, any> = {
      model: this.config.model,
      messages: prepMessages(messages),
      temperature: opts?.temperature ?? this.defaultTemperature,
      stream: false,
    }

    if (tools.length > 0) {
      body.tools = tools
      body.tool_choice = 'auto'
    }

    const data = await this.request(body, false)
    const choice = data.choices?.[0]
    if (!choice) throw new Error('LLM 返回空响应')

    const { content, thinking } = this.extractThinking(choice)

    let toolCalls: ToolCall[] | null = null
    if (choice.message.tool_calls && choice.message.tool_calls.length > 0) {
      toolCalls = choice.message.tool_calls.map((tc: any) => {
        let args: Record<string, any> = {}
        if (tc.function.arguments) {
          try {
            args =
              typeof tc.function.arguments === 'string'
                ? JSON.parse(tc.function.arguments)
                : tc.function.arguments
          } catch {
            args = { _raw: tc.function.arguments }
          }
        }
        return {
          id: tc.id,
          type: 'function' as const,
          function: {
            name: tc.function.name,
            arguments: args,
          },
        }
      })
    }

    return {
      message: {
        role: choice.message.role,
        content,
        thinking,
        tool_calls: toolCalls,
      },
    }
  }

  // ── checkHealth() ──────────────────────────────────────────────────────

  async checkHealth(): Promise<HealthStatus> {
    const status: HealthStatus = { apiOk: false, modelOk: false }

    if (!this.isConfigured) return status

    try {
      // Minimal request to verify API key and model
      const body = {
        model: this.config.model,
        messages: [{ role: 'user', content: 'hi' }],
        max_tokens: 1,
        stream: false,
      }

      const response = await fetch(this.endpoint, {
        method: 'POST',
        headers: this.buildHeaders(),
        body: JSON.stringify(body),
      })

      status.apiOk = response.status !== 401 && response.status !== 403
      status.modelOk = response.ok

      return status
    } catch {
      return status
    }
  }
}

export const llmClient = new LLMClient()
